groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+template <size_t INDEX, class HEAD = VARIANT_NOT_A_TYPE, class... TAIL>
+struct Variant_VariantAlternativeImpl
+: Variant_VariantAlternativeImpl<INDEX - 1, TAIL...> {
+};
+
+template <size_t INDEX>
+struct Variant_VariantAlternativeImpl<INDEX, VARIANT_NOT_A_TYPE> {
+};
+
+template <class HEAD, class... TAIL>
+struct Variant_VariantAlternativeImpl<0, HEAD, TAIL...> {
+    typedef HEAD type;
+};
+
+template <size_t INDEX, class HEAD, class... TAIL>
+struct variant_alternative<INDEX, variant<HEAD, TAIL...> >
+: Variant_VariantAlternativeImpl<INDEX, HEAD, TAIL...> {
+    // This metafunction calculates the alternative type with index 'INDEX' in
+    // the supplied 'bsl::variant' type.
+};
+
+#endif
+
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_ALIAS_TEMPLATES
+template <size_t INDEX, class TYPE>
+using variant_alternative_t = typename variant_alternative<INDEX, TYPE>::type;
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_ALIAS_TEMPLATES
+
+template <class BSL_VARIANT>
+struct variant_size;  // primary template not defined
+
+// The following metafunctions calculate the size of a cv qualified 'variant'
+// type.
+template <class BSL_VARIANT>
+struct variant_size<const BSL_VARIANT> : variant_size<BSL_VARIANT> {
+};
+
+template <class BSL_VARIANT>
+struct variant_size<volatile BSL_VARIANT> : variant_size<BSL_VARIANT> {
+};
+
+template <class BSL_VARIANT>
+struct variant_size<const volatile BSL_VARIANT> : variant_size<BSL_VARIANT> {
+};
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+template <class HEAD, class... TAIL>
+struct variant_size<variant<HEAD, TAIL...> >
+: bsl::integral_constant<size_t, 1 + sizeof...(TAIL)> {
+    // This metafunction calculates the number of alternatives in a given
+    // 'variant' type.
+};
+#endif
+
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_VARIABLE_TEMPLATES
+template <typename BSL_VARIANT>
+inline
+constexpr size_t variant_size_v = variant_size<BSL_VARIANT>::value;
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_VARIABLE_TEMPLATES
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+// FREE FUNCTIONS
+template <class HEAD, class... TAIL>
+void swap(bsl::variant<HEAD, TAIL...>& lhs, bsl::variant<HEAD, TAIL...>& rhs);
+    // Efficiently exchange the values of the specified 'lhs' and 'rhs'
+    // objects.  This method provides the no-throw exception-safety guarantee
+    // if  TODO
+
+// HASH SPECIALIZATIONS
+template <class HASHALG, class HEAD, class... TAIL>
+void hashAppend(HASHALG& hashAlg, const variant<HEAD, TAIL...>& input);
+    // Pass the specified 'input' to the specified 'hashAlg', where 'hashAlg'
+    // is a hashing algorithm.
+
+// 20.7.5, value access
+template <class TYPE, class HEAD, class... TAIL>
+bool holds_alternative(const variant<HEAD, TAIL...>&) BSLS_KEYWORD_NOEXCEPT;
+    // Return 'true' if the supplied 'variant' objects currently holds the
+    // 'TYPE' alternative, and false otherwise.  This function is defined only
+    // if 'TYPE' appears exactly once in the variant's list of alternatives.
```

Comment:
It's not accurate to say that the function is defined only if `TYPE` appears exactly once in the variant's list of alternatives. It is defined even if that's not the case, but the call is ill-formed (at least assuming that we always check the condition; see my comment on the definition.)
```suggestion
bool holds_alternative(
                      const variant<HEAD, TAIL...>& obj) BSLS_KEYWORD_NOEXCEPT;
    // Return 'true' if the specified 'obj' currently holds the 'TYPE'
    // alternative, and 'false' otherwise.  'TYPE' shall appear exactly once in
    // the variant's list of alternatives.
```